add resource filter handling before printing

juanvallejo opened the pull request:
**Release note**:

``` release-note
release-note-none
```

Resources are currently filtered (in order to prevent printing) at print
time in their HumanReadablePrinter handlers. This design makes it not
possible to filter objects when they are printed using any other
printer, such as YAML, JSON, or the NamePrinter.

This patch removes any filters previously added at the printer level for
pods and adds a way to define resource-specific filters before they are
sent to a printer handler. A woking filter handler for pods has also been
implemented.

Filters affect resources being printed through the HumanReadablePrinter,
YAML, JSON, and `--template` printers.

cc @smarterclayton 

<!-- Reviewable:start -->

---

This change is [<img src="https://reviewable.kubernetes.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.kubernetes.io/reviews/kubernetes/kubernetes/31163)

<!-- Reviewable:end -->

brendandburns commented on pkg/kubectl/cmd/get.go:
```diff
@@ -296,26 +291,24 @@ func RunGet(f *cmdutil.Factory, out io.Writer, errOut io.Writer, cmd *cobra.Comm
-		if err := printer.PrintObj(obj, out); err != nil {
-			allErrs = append(allErrs, err)
+		for ix := range infos {
+			filter := kubectl.NewResourceFilter(filterOptionsForCommand(cmd, allNamespaces))
```
Why are you creating this inside the for loop?

brendandburns commented on pkg/kubectl/cmd/get.go:
```diff
@@ -296,26 +291,24 @@ func RunGet(f *cmdutil.Factory, out io.Writer, errOut io.Writer, cmd *cobra.Comm
-			allErrs = append(allErrs, err)
+		for ix := range infos {
+			filter := kubectl.NewResourceFilter(filterOptionsForCommand(cmd, allNamespaces))
+			if isFiltered, _ := filter.Filter(infos[ix].Object); !isFiltered {
```
Handle the error here.

brendandburns commented on pkg/kubectl/cmd/get.go:
```diff
@@ -391,6 +386,14 @@ func RunGet(f *cmdutil.Factory, out io.Writer, errOut io.Writer, cmd *cobra.Comm
+
+		// filter objects if filter has been defined for current object
+		filter := kubectl.NewResourceFilter(filterOptionsForCommand(cmd, allNamespaces))
+		if isFiltered, _ := filter.Filter(original); isFiltered {
```
Handle the error here.

brendandburns commented on pkg/kubectl/cmd/get.go:
```diff
@@ -451,3 +454,23 @@ func mustPrintWithKinds(objs []runtime.Object, infos []*resource.Info, sorter *k
+// implement specific filter handlers for resources before they are printed
+func filterOptionsForCommand(cmd *cobra.Command, withNamespace bool) *kubectl.FilterOptions {
+	columnLabel, err := cmd.Flags().GetStringSlice("label-columns")
+	if err != nil {
```
log a warning or error.

brendandburns commented on pkg/kubectl/resource_filter.go:
```diff
@@ -0,0 +1,100 @@
+}
+
+// FilterOptions implements PrintOptions
+type FilterOptions struct {
```
Why introduce a separate object for this?

smarterclayton replied:
Agree, we don't need ResourceFilter either.

Filter should accept a type of `type FilterFunc func(runtime.Object, PrintOptions) (bool, error)`.   Simply define a function `FilterDeletedPods` and then have `kubectl.Factory` have a method for `DefaultResourceFilters` that returns an array of `FilterFunc`

brendandburns commented on pkg/kubectl/resource_filter.go:
```diff
@@ -0,0 +1,100 @@
+	}
+	funcType := filterFunc.Type()
+	if funcType.NumIn() != 2 || funcType.NumOut() != 1 {
+		return fmt.Errorf("invalid filter handler." +
```
Print the wrong values (e.g. "Must have 2 parameters, actual <foo>")

stevekuznetsov commented on pkg/kubectl/resource_printer.go:
```diff
@@ -446,9 +445,9 @@ func (h *HumanReadablePrinter) HandledResources() []string {
-func (h *HumanReadablePrinter) FinishPrint(output io.Writer, res string) error {
-	if !h.options.NoHeaders && !h.options.ShowAll && h.hiddenObjNum > 0 {
-		_, err := fmt.Fprintf(output, "  info: %d completed object(s) was(were) not shown in %s list. Pass --show-all to see all objects.\n\n", h.hiddenObjNum, res)
+func (h *HumanReadablePrinter) FinishPrint(output io.Writer, res string, hiddenObjNum int) error {
```
Seems like a really high cost (and a lot of literal `0` getting passed into `FinishPrint()` impls for this -- maybe there's a way to do this with less effort? The Printers no longer should care about filtering and this keeps a tenuous connection there

juanvallejo replied:
Yeah, definitely agree. I will look into making this part of the filter, rather than leaving it as part of the printer

juanvallejo commented:
@brendandburns @smarterclayton PTAL

> Agree, we don't need ResourceFilter either.
>
> Filter should accept a type of type FilterFunc func(runtime.Object, PrintOptions) (bool, error). Simply define a function FilterDeletedPods and then have kubectl.Factory have a method for DefaultResourceFilters that returns an array of FilterFunc

Hm, I looked into making `FilterOptions` a type of `FilterFunc` array, however, when I receive these in `cmd/get.go`, wouldn't I have to iterate through each `FilterFunc` inside of the loop iterating through resources?

From https://github.com/openshift/origin/pull/10574#discussion_r75754869:
> This should be in pkg/kubectl/resource

Since the filter package / each individual filter function requires access to `PrintOptions` (defined in `kubectl`), and `cmd/get.go` already imports `kubectl` when calling `Filter...`, I cannot think of a way of doing this without causing a dependency cycle.

stevekuznetsov commented on pkg/kubectl/cmd/get.go:
```diff
@@ -296,26 +292,24 @@ func RunGet(f *cmdutil.Factory, out io.Writer, errOut io.Writer, cmd *cobra.Comm
-			return err
+		for ix := range infos {
+			if isFiltered, err := filter.Filter(infos[ix].Object); !isFiltered {
+				if err != nil {
```
What are the implications of ignoring the error if it is filtered?

juanvallejo replied:
If `isFiltered == true` the error returned is simply "error: no filter type for <resource_type>", while not something to output here, I could glog it

smarterclayton commented:
Moving this to 1.5 given timing.

juanvallejo commented:
@smarterclayton or @deads2k PTAL: https://github.com/kubernetes/kubernetes/pull/31163/commits/3deecaf421b92d04694bf83e9e907d77ebaca066

brendandburns commented on pkg/kubectl/resource_filter.go:
```diff
@@ -0,0 +1,113 @@
+	if pod.Status.Reason != "" {
+		reason = pod.Status.Reason
+	}
+	if !options.ShowAll && (reason == string(api.PodSucceeded) || reason == string(api.PodFailed)) {
```
nit: 
```go
return !options.ShowAll && ...
```

brendandburns commented on pkg/kubectl/cmd/get.go:
```diff
@@ -20,8 +20,12 @@ import (
 	"io"
 
+	"github.com/golang/glog"
+
```
nit: delete empty line here.

brendandburns commented on pkg/kubectl/cmd/get.go:
```diff
@@ -20,8 +20,12 @@ import (
+	"github.com/golang/glog"
+
 	"github.com/renstrom/dedent"
+
```
here too.

brendandburns commented on pkg/kubectl/cmd/get.go:
```diff
@@ -20,8 +20,12 @@ import (
 	"github.com/renstrom/dedent"
+
 	"github.com/spf13/cobra"
+
```
and here.

brendandburns commented on pkg/kubectl/cmd/get.go:
```diff
@@ -296,26 +296,26 @@ func RunGet(f *cmdutil.Factory, out io.Writer, errOut io.Writer, cmd *cobra.Comm
 			res = infos[0].ResourceMapping().Resource
 		}
 
-		obj, err := resource.AsVersionedObject(infos, !singular, version, f.JSONEncoder())
```
Why did you drop AsVersionedObject for the infos?

juanvallejo replied:
Each versionedObject would have a different `reflect.Type` than [generic `infos.Object` in HumanReadablePrinter](https://github.com/kubernetes/kubernetes/blob/master/pkg/kubectl/cmd/get.go#L380), so when filtering I would have something like `*api.Pod` [as the map key for the pod filter](https://github.com/kubernetes/kubernetes/pull/31163/commits/5de8f7ed7f5a9502666d6a6008c432587bb3f19a#diff-baa0ea36296411cb41a034c9c5f6ec13R44), which would not match with the `reflect.Type` of versioned objects.

A solution that would allow me to keep versionedObjects here would be to have two keys (one for versioned, another for unversioned) in the [filterMap](https://github.com/kubernetes/kubernetes/pull/31163/commits/5de8f7ed7f5a9502666d6a6008c432587bb3f19a#diff-baa0ea36296411cb41a034c9c5f6ec13R28) pointing to the same filterFunc for that resource. WDYT?

brendandburns commented on pkg/kubectl/resource_filter.go:
```diff
@@ -0,0 +1,110 @@
+
+type FilterFunc func(runtime.Object) (bool, error)
+
+type FilterOptions struct {
```
godoc this type.

brendandburns replied:
also, it's weird that this is a public type but only has private members.  If you really want to keep the implementation private, I would add an interface for the public functions and then have the struct implement the interface.

brendandburns commented on pkg/kubectl/resource_filter.go:
```diff
@@ -0,0 +1,110 @@
+
+// Handler adds a filter handler to a FilterOptions instance.
+// See validateFilterFunc for required method signature.
+func (f *FilterOptions) Handler(handlerFn interface{}) error {
```
Don't really like `Handler` as a name.  `AddFilter`?

Also, why are you taking an interface and then doing validation, why not just explicitly take a `FilterFunc`?

brendandburns commented on pkg/kubectl/resource_filter.go:
```diff
@@ -0,0 +1,110 @@
+type FilterFunc func(runtime.Object) (bool, error)
+
+type FilterOptions struct {
+	filterMap    map[reflect.Type]reflect.Value
```
Why are you keeping this a `reflect.Value` don't you know that it is a `FilterFunc` here?

juanvallejo commented:
@brendandburns 
> Why did you drop AsVersionedObject for the infos?

Updated filter to [handle versioned objects](https://github.com/kubernetes/kubernetes/pull/31163/commits/63616c4ed3c183d16261512c8531515a80571797). PTAL

juanvallejo commented:
@smarterclayton 
added handling for versioned objects [here](https://github.com/kubernetes/kubernetes/pull/31163/commits/25478964ffb2bdb7296ddc1b242b0581a44b7b70#diff-baa0ea36296411cb41a034c9c5f6ec13R46), WDYT?

brendandburns commented on pkg/kubectl/resource_filter.go:
```diff
@@ -0,0 +1,107 @@
+package kubectl
```
needs boilerplate.

juanvallejo replied:
Done! https://github.com/kubernetes/kubernetes/pull/31163/commits/928453f22b3331012f370b402491900380761e88

brendandburns commented on pkg/kubectl/resource_filter.go:
```diff
@@ -0,0 +1,107 @@
+// a map of FilterFuncs for given resource types, PrintOptions for a
+// resource, and a hiddenObjNum to keep track of the number of filtered resources
+type filterOptions struct {
+	filterMap    map[reflect.Type]FilterFunc
```
why is the key `reflect.Type` and not `runtime.Object`?

juanvallejo replied:
It's so that I can tie the FilterFunc to a resource type, rather than the entire object. Also, if I use `runtime.Object` as the key there would be no way for me to `AddFilter` since I would need the value of the object beforehand

brendandburns commented on pkg/kubectl/resource_filter.go:
```diff
@@ -0,0 +1,107 @@
+	f.AddFilter(reflect.TypeOf(&v1.Pod{}), filterPods)
+}
+
+func (f *filterOptions) AddFilter(objType reflect.Type, handlerFn FilterFunc) error {
```
why not just take the `runtime.Object` interface here as the first parameter?

brendandburns commented on pkg/kubectl/resource_filter.go:
```diff
@@ -0,0 +1,107 @@
+
+	switch p := obj.(type) {
+	case *v1.Pod:
+		reason = string(p.Status.Phase)
```
this code is duplicated, is go smart enough to let you join the cases together?

juanvallejo replied:
I tried doing a `fallthrough`, but it apparently cannot be done on a type switch [[1]](http://stackoverflow.com/questions/11531264/why-isnt-fallthrough-allowed-in-a-type-switch)[[2]](http://stackoverflow.com/a/18344196)

A solution to minimize code reuse could be to have a second function that extracts a pod's `Status.Phase` string, and have this function called from each switch case:

```Go
case *v1.Pod:
		reason = getPodReason(p)
case *api.Pod:
		reason = getPodReason(p)
```

WDYT?

brendandburns commented on pkg/kubectl/resource_filter.go:
```diff
@@ -0,0 +1,107 @@
+		return false
+	}
+
+	return !options.ShowAll && (reason == string(api.PodSucceeded) || reason == string(api.PodFailed))
```
You're potentially comparing against `api` and `v1` here, while that works for now, it's not guaranteed to work in the future, if we ever change `api.PodSucceeded` (e.g. in `v2`)

brendandburns commented on pkg/kubectl/resource_filter.go:
```diff
@@ -0,0 +1,107 @@
+	t := reflect.TypeOf(obj)
+	if filter, ok := f.filterMap[t]; ok {
+		args := []reflect.Value{reflect.ValueOf(obj), reflect.ValueOf(f.options)}
+		filterFunc := reflect.ValueOf(filter)
```
I'm pretty sure this reflection is unnecessary:

```go
if filter, found := f.filterMap[t]; found { 
   isFiltered := filter(obj, f.options)
   ...
```

smarterclayton commented on pkg/kubectl/resource_filter.go:
```diff
@@ -0,0 +1,115 @@
+// ResourceFilter filters api resources
+type ResourceFilter interface {
+	Filter(obj runtime.Object) (bool, error)
+	AddFilter(reflect.Type, FilterFunc) error
```
I don't see the need for this abstraction.  It would be better if you simply accepted a list of functions, and functions did type casts to see if they recognized the type.  Reflect and maps are very slow - a simple `if obj, ok := input.(MyType); ok` will be 2-3 orders of magnitude faster and avoid the need for reflection.

If you find yourself needing to change the list of filters, just wrap them.  You shouldn't need to have a mutator on the filter in order to be able to accomplish your goal:

```
type Filters []Filterer

func (f Filters) Filter(obj runtime.Object) (bool, error) {
  for _, filter := range f {
    ok, err := filter.Filter(obj)
    if err != nil {
      return false, err
    }
    if ok {
      return true, nil
    }
  }
  return false, nil
}
```

Is much much simpler.  Filter count accumulation doesn't belong in this interface - it belongs in the thing calling Filter.

smarterclayton replied:
If you need to have a helper to filter a list of []runtime.Object, do that instead and have it return the number of omitted, an updated []runtime.Object, or an error.

And really this should be using the visitor pattern we already have in place, but we can come back to that later (visitor should be able to handle this transparently today - it's only because we've accumulated a layer of hacks on get.go that it's not as clean today).

smarterclayton commented on pkg/kubectl/resource_filter.go:
```diff
@@ -0,0 +1,115 @@
+
+// Filter extracts the filter handler, if one exists, for the given resource
+func (f *filterOptions) Filter(obj runtime.Object) (bool, error) {
+	t := reflect.TypeOf(obj)
```
I don't think you need to use reflection in this context.  Just use a type function.

brendandburns commented on pkg/kubectl/cmd/util/helpers.go:
```diff
@@ -631,3 +632,66 @@ func MaybeConvertObject(obj runtime.Object, gv unversioned.GroupVersion, convert
+
+// FilterResourceList receives a list of runtime objects.
+// If any objects are filtered, that number is returned along with a modified list.
+func FilterResourceList(obj runtime.Object, filterFuncs kubectl.Filters, filterOpts *kubectl.PrintOptions) (int, []runtime.Object, []error) {
```
`ErrorList` instead of `[]error` or even better use `NewAggregate` to produce a single error.

juanvallejo replied:
Thanks for the feedback, I have updated the function to return an Aggregate error: https://github.com/kubernetes/kubernetes/pull/31163/commits/1600392ac75a6a207b2ad034392fcb56a9f1064d

I can go ahead and squash if that is okay

brendandburns commented:
Generally LGTM, one small-ish nit.

brendandburns commented:
@k8s-bot ok to test

brendandburns reviewed (changes requested):
one small-ish nit.

brendandburns commented:
Happy to see a squash, and check the tests which appear to be failing.

juanvallejo commented:
@k8s-bot gce e2e test this

juanvallejo commented on pkg/kubectl/cmd/util/helpers.go:
```diff
@@ -626,3 +627,78 @@ func MaybeConvertObject(obj runtime.Object, gv unversioned.GroupVersion, convert
+
+// ObjectListToVersionedObject receives a list of api objects and a group version
+// and squashes the list's items into a single versioned runtime.Object.
+func ObjectListToVersionedObject(objects []runtime.Object, version unversioned.GroupVersion) (runtime.Object, error) {
```
@brendandburns Turns out that the reason some unit tests were not passing was due to a `runtime.Decode` call [in the `get` cmd unit test](https://github.com/kubernetes/kubernetes/blob/master/pkg/kubectl/cmd/get_test.go#L568) which was expecting a single output from the printer, rather than multiple calls to the printer when iterating through a filtered `[]runtime.Object` list. Because of this, I defined this helper function in order to turn a `[]runtime.Object` list into a single versioned `runtime.Object`. Please take one more look before squashing.

juanvallejo committed:
add resource handling before printing

Resources are currently filtered (in order to prevent printing) at print
time in their HumanReadablePrinter handlers. This design makes it not
possible to filter objects when they are printed using any other
printer, such as YAML, JSON, or the NamePrinter.

This patch removes any filters previously added at the printer level for
pods and adds a way to define resource-specific filters before they are
sent to a printer handler. A woking filter handler for pods has also
been
implemented.

Filters affect resources being printed through the HumanReadablePrinter,
YAML, JSON, and `--template` printers.

brendandburns commented:
@k8s-bot test this please issue: #31776

brendandburns commented:
LGTM.
